### ISSUE TITLE
Enchancement: [strict-boolean-expressions] Treat `!!x` same as `Boolean(x)`

### ISSUE THREAD

**Issue by Faithfinder** [typescript-eslint/typescript-eslint#9060]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&fileType=.tsx&code=GYVwdgxgLglg9mABACwKYBt1wBQFsCeAbgIYBOAXIgM5SkxgDmiAPomCJgJSIDeAUIkQxgibAEIxBEqW49EAXz6KgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs%2BSAIZN6AD2LQUySnyaoMkRNGh9okcGAC%2BILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function hello(myvar: string | null) {
  if (!!myvar) { }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/strict-boolean-expressions": ["error"],
  },
};
```

### Additional Info

Not strictly a bug, but IMO this is a valid, less verbose way to explicitly convert things to boolean.

**Comment by auvred** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/7444, https://github.com/typescript-eslint/typescript-eslint/issues/3470

**Comment by Faithfinder** [comment]
Weird, I searched by rule name and found nothing, sorry.

Still, in #7444 the author's point isn't addressed and I'd like to second it. `Boolean()` is in the proposed fixes list, and `!!` is equivalent. Either remove `Boolean()` as a solution, or allow `!!`?

https://github.com/typescript-eslint/typescript-eslint/issues/7444#issuecomment-1671829329

**Comment by Faithfinder** [comment]
I'm still getting a feel for the rule. I'm in a new job and there was neither linting, nor Typescript here before me. Previously I've used _mostly_ the recommended config, and learned that some fairly useful and obvious things are missing from it, so this time I've explored every single rule in eslint, typescript-eslint and react plugins, enabling those that looked reasonable to me. I've hit this one in the first file conversion to Typescript I've made. In that case, I've opted into doing a `!== null` check instead of `!!`. I'm not sure I won't disable this rule, or relax it as much as possible in the future.

Still, IMO, allowing `Boolean()` as a bypass, and forbidding `!!` is just weird. Now, I don't know if what you're saying about it being uncommon syntax is true, but it was widely used by my previous team, there's 119 hits in the JS codebase of my current one, and [56](https://github.com/search?q=repo%3Atypescript-eslint%2Ftypescript-eslint%20!!&type=code) on a naive search of this repo right here. It very clear tells me this is an explicit coercion.

I don't _really_ care whether you remove Boolean bypass or allow `!!` though, I do strongly believe they should be treated the same, though.

**Comment by bradzacher** [comment]
Sorry - I should clarify.
"In codebases that use `strict-boolean-expressions` usage of `!!` to loosely coerce a value to a boolean is uncommon"

The syntax itself is reasonably common but when people opt-in to the SBE rule they explicitly want to NOT have weak coercion of values - thus it's rare that they use the rule and also want to allow the syntax.

> Still, IMO, allowing Boolean() as a bypass, and forbidding !! is just weird.

There are multiple reasons that it works this way.

The first (and main) reason is that the argument passed to `Boolean()` is not a boolean context - it's a function that accepts `any` and returns a boolean. So it's not explicitly handled by any of the rule's logic by default as the rule was designed to inspect boolean contexts aka logical expressions.

OTOH `!value` is a logical expression that the rule checks intentional and by design. So following from that `!!value` is also checked by design as `!!value` is no different to `!value` from the rule's perspective.

The reason we haven't made a change to specifically allow the `!!` case is as mentioned - it's very uncommon people actually want to allow it when using the rule - as evidenced by the fact that 

The difference is the verbosity of it.
`!!value`
vs
`Boolean(value)`

The former is very easy to misread as `!value` and miss the fact that it's in-fact doing a weak coercion.

I could see the possibility of an option for it - but as per the above - I don't think it's broadly applicable enough for us to move and implement it in our project.

**Comment by Faithfinder** [comment]
I see. Well, thank you for the explanation, it's very nice of you to go into such detail :) I still think they should match regardless, but at least it doesn't look like a missed case anymore. Nothing more I can say, so, see you around :)
